cpp/src/structure/remove_self_loops_impl.cuh

```diff
@@ -0,0 +1,180 @@
+/*
+ * Copyright (c) 2023, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <raft/core/handle.hpp>
+#include <rmm/device_uvector.hpp>
+
+#include <thrust/distance.h>
+#include <thrust/iterator/zip_iterator.h>
+#include <thrust/remove.h>
+#include <thrust/tuple.h>
+
+#include <algorithm>
+#include <optional>
+
+namespace cugraph {
+
+namespace detail {
+
+template <typename vertex_t>
+void remove_self_loops(raft::handle_t const& handle,
+                       rmm::device_uvector<vertex_t>& edgelist_srcs /* [INOUT] */,
+                       rmm::device_uvector<vertex_t>& edgelist_dsts /* [INOUT] */)
+{
+  auto edge_first =
+    thrust::make_zip_iterator(thrust::make_tuple(edgelist_srcs.begin(), edgelist_dsts.begin()));
+  edgelist_srcs.resize(
+    thrust::distance(
+      edge_first,
+      thrust::remove_if(handle.get_thrust_policy(),
+                        edge_first,
+                        edge_first + edgelist_srcs.size(),
+                        [] __device__(auto e) { return thrust::get<0>(e) == thrust::get<1>(e); })),
+    handle.get_stream());
+  edgelist_dsts.resize(edgelist_srcs.size(), handle.get_stream());
+}
+
+template <typename vertex_t, typename A>
+void remove_self_loops(raft::handle_t const& handle,
+                       rmm::device_uvector<vertex_t>& edgelist_srcs /* [INOUT] */,
+                       rmm::device_uvector<vertex_t>& edgelist_dsts /* [INOUT] */,
+                       rmm::device_uvector<A>& edgelist_a /* [INOUT] */)
```

Comment:
I am not sure we need three separate functions that takes 1, 2, or 3 additional variables in addition to edgelist_srcs & edgelist_dsts.

See https://github.com/rapidsai/cugraph/blob/branch-23.12/cpp/src/structure/detail/structure_utils.cuh#L148

Then, we can set

edge_first = thrust::make_zip_iterator(edgelist_srcs.begin(), edgelist_dsts.begin(), get_dataframe_buffer_begin());

And the remaining code will be same. Same for the sort_and_remove_multi_edges function.

This will be more important if we bring additional code to limit peak memory footprint.